tests/performance/density.go

```diff
@@ -0,0 +1,316 @@
+
+package performance
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+
+	kvv1 "kubevirt.io/client-go/api/v1"
+	cd "kubevirt.io/kubevirt/tests/containerdisk"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests"
+	"kubevirt.io/kubevirt/tests/util"
+
+	e2emetrics "kubevirt.io/kubevirt/tests/performance/metrics"
+)
+
+type DensityTest struct {
+	NumVMs               int               `yaml:"num_vms"`                 // number of vms
+	ArrivalRate          float64           `yaml:"arrival_rate"`            // number fo vms per second
+	ArrivalUnit          time.Duration     `yaml:"arrival_unit"`            // to convert from second to another unit
+	ArrivalRateFuncName  string            `yaml:"arrival_rate_func_name"`  // e.g. using a constant interval or a poisson process
+	VMImage              cd.ContainerDisk  `yaml:"vm_image_name"`           // the image type will reflect in the storage size
+	CPULimit             string            `yaml:"cpu_limit"`               // number of CPUs to allocate (100m = .1 cores)
+	MEMLimit             string            `yaml:"mem_limit"`               // amount of Memory to allocate
+	VMStartupLimits      e2emetrics.Metric `yaml:"vm_startup_limits"`       // percentile limit of single vm startup latency
+	VMBatchStartupLimit  time.Duration     `yaml:"vm_batch_startup_limit"`  // upbound of startup latency of a batch of vms
+	VMDeletionLimit      time.Duration     `yaml:"vm_deletion_limit"`       // time limit of single vm deletion in seconds
+	VMBatchDeletionLimit time.Duration     `yaml:"vm_batch_deletion_limit"` // time limit of a batch of vms deletion
+	CloudInitUserData    string            `yaml:"cloud_init_user_data"`    // script to run into the VM during the startup
+	GinkoLabel           string            `yaml:"ginko_label"`             // test label [small-scale] [medium-scale] [large-scale]
+}
+
+// ArrivalRateFunc generate arrival time. The default is a stepped load with an interval of 300 miliseconds
+func (t *DensityTest) ArrivalRateFunc(arrivalRate float64) time.Duration {
+	if strings.Contains(string(t.ArrivalRateFuncName), "poisson") {
+		return poissonProcess(arrivalRate)
+	}
+	// ArrivalRate is the number fo vms per second. Then, we calculate a stepped interval between VMs
+	if t.ArrivalRate > 0 {
+		return time.Duration(1/t.ArrivalRate) * time.Millisecond
+	}
+	return time.Duration(300) * time.Millisecond
+}
+
+var _ = SIGDescribe("Control Plane Performance Density Testing", func() {
+	var (
+		err        error
+		virtClient kubecli.KubevirtClient
+	)
+
+	BeforeEach(func() {
+		virtClient, err = kubecli.GetKubevirtClient()
+		util.PanicOnError(err)
+		tests.BeforeTestCleanup()
+	})
+
+	densityTests := []DensityTest{
+		{
+			NumVMs:              100,
+			ArrivalRate:         0.002, // Vms per second
```

Comment:
I don't fully understand this. Why `per second` if the `ArrivalUnit` is set to `time.Millisecond`?